parquet-column/src/main/java/org/apache/parquet/filter2/predicate/FilterApi.java

```diff
@@ -257,6 +266,16 @@ public static <T extends Comparable<T>, C extends Column<T> & SupportsEqNotEq> N
     return new NotIn<>(column, values);
   }
 
+  public static <T extends Comparable<T>, C extends Column<T> & SupportsContains> Contains<T> contains(
```

Comment:
Thinking to the future, it could eventually be useful to support Array FilterPredicates like "repeated field X contains an int field greater than Y". We could refactor this API to be written like:

```java
FilterApi.contains(
  FilterApi.eq(
    FilterApi.arrayColumn(FilterApi.intColumn("repeated_int_field")), 
    100
  )
)
```

& we could eventually support `FilterApi.contains(FilterApi.lt(...))`, `FilterApi.contains(FilterApi.gt(...))`, etc...

Reply:
It's not a standard SQL function, but I've seen it in SQL extension languages such as [BigQuery Standard SQL](https://cloud.google.com/bigquery/docs/reference/standard-sql/functions-and-operators#in_operators), and I've gotten several requests to support this by users of the [Scio](https://github.com/spotify/scio) Parquet library!

that's a good point about making this composable, I think it would be more efficient to do `CONTAINS(a or b)` than `CONTAINS(a) or CONTAINS(b)`. What do you think about supporting `lt/gt` in addition to `eq`-based Contains?  for example, `CONTAINS(eq(a) OR gt(b))` ? It would make this PR a lot more complex but I'm happy to try. We could probably re-use a lot of the existing filter code for `eq`, `lt/gt`, etc...

Reply:
Yes, I agree that it worths making the function composable. Perhaps we can define all the necessary API (with implementation of the most common ones) in the PR and implementations of all other kinds of inputs can be split into separate PRs. Let's not hurry for the coming release. It still takes time to stabilize.

Reply:
Sounds great. I would love to get predicates for array size in there, too, but I think that one is going to be a bit more complex (we might need to update Statistics to reflect array sizes?)... that one can wait for a separate PR.

Reply:
Actually, I think I need some clarification on the expected behavior of And/Or for an array field...

Say we have the following records, each containing a repeated int field with the following values:

```
Record1:
  repeated_int_field: [2, 3, 4, 5]
Record2:
  repeated_int_field: [1, 7]
```

Given the predicate `contains(and(gt(3), lt(6)))`, should we return:

1. Only Record1, because it contains **at least one element that satisfies both predicates** gt(3) and lt(6), or
2. Both Record1 and Record2, because Record2 contains **elements that satisfy each predicate separately** (7 matches `gt(3)` but not `lt(6)`, and 1 satisfies `lt(6)` but not `gt(3)`?

I think I would assume 2 🤔 but not totally sure. 

cc @gszadovszky as well :)

Reply:
I would expect the option 1. Option 2 looks more like contains(or(gt(3),lt(6))). The case would be more complex if we add NOT expr to it. If there is any confusion like this, what about only supporting eq and ne for now? More complicated expressions need more research and inputs from other experts.

Reply:
That sounds good! 👍

Reply:
Ok, I've pushed my changes to support composable contains predicates! example usage:

```java
FilterApi.containsOr(
  FilterApi.containsEq(longColumn("phoneNumbers.phone.number"), 5555555555L),
  FilterApi.containsOr(
    FilterApi.containsEq(longColumn("phoneNumbers.phone.number"), -10000000L), // Won't be matched
    FilterApi.containsEq(longColumn("phoneNumbers.phone.number"), 2222222222L)));
```

I left out implementing DoesNotContain/ContainsNotEq for now to keep the PR simpler to parse. Let me know if the API looks ok 👍 Then I can fill out the rest of the implementations/make the unit tests more thorough.

Reply:
> It's not a standard SQL function, but I've seen it in SQL extension languages such as [BigQuery Standard SQL](https://cloud.google.com/bigquery/docs/reference/standard-sql/functions-and-operators#in_operators), and I've gotten several requests to support this by users of the [Scio](https://github.com/spotify/scio) Parquet library!
> 
> that's a good point about making this composable, I think it would be more efficient to do `CONTAINS(a or b)` than `CONTAINS(a) or CONTAINS(b)`. What do you think about supporting `lt/gt` in addition to `eq`-based Contains? for example, `CONTAINS(eq(a) OR gt(b))` ? It would make this PR a lot more complex but I'm happy to try. We could probably re-use a lot of the existing filter code for `eq`, `lt/gt`, etc...

I took a second look, it seems that you were mentioning `IN` expression which we already have: https://github.com/apache/parquet-mr/blob/master/parquet-column/src/main/java/org/apache/parquet/filter2/predicate/FilterApi.java#L209-L258

Reply:
> Ok, I've pushed my changes to support composable contains predicates! example usage:
> 
> ```java
> FilterApi.containsOr(
>   FilterApi.containsEq(longColumn("phoneNumbers.phone.number"), 5555555555L),
>   FilterApi.containsOr(
>     FilterApi.containsEq(longColumn("phoneNumbers.phone.number"), -10000000L),
>     FilterApi.containsEq(longColumn("phoneNumbers.phone.number"), 2222222222L)));
> ```
> 
> I left out implementing DoesNotContain/ContainsNotEq for now to keep the PR simpler to parse. Let me know if the API looks ok 👍 Then I can fill out the rest of the implementations/make the unit tests more thorough.

Sorry I didn't make it clear. I thought we could still leverage [existing composite expression](https://github.com/apache/parquet-mr/blob/master/parquet-column/src/main/java/org/apache/parquet/filter2/predicate/FilterApi.java#L293-L326) and combine `CONTAINS(x)` sub-expressions under the same level of `AND` or `OR` expression during rewrite process like https://github.com/apache/parquet-mr/blob/master/parquet-column/src/main/java/org/apache/parquet/filter2/compat/FilterCompat.java#L79. It would be harder to maintain the code if introduce new specialized composite expression like `containsOr`.

Reply:
ahh, ok. So the rewriter would inspect the runtime classes of the `left` and `right` predicates of an `And` or `Or`, and if they match the new Contains* expressions, we rewrite it into a single `Contains` predicate containing both clauses?

Reply:
I would expect so, but it seems requiring a lot more work. We can support the simplest case at this point.

Reply:
ok, I think I mostly got this working, supporting just Contains Eq (+ composition using And/Or) to start. hopefully this is closer to what you had in mind!